examples/triangle_intersection/triangle_intersection.cpp

```diff
@@ -0,0 +1,384 @@
+  TriangleIntersectionCallback(
+      Triangles<DeviceType> triangles, Kokkos::View<int *, DeviceType> offsets,
+      Kokkos::View<ArborX::Point *, DeviceType> coefficients)
+      : triangles_(triangles)
+      , offsets_(offsets)
+      , coefficients_(coefficients)
+  {}
+
+  // The search tree consists entirely of boxes, although the primitives are
+  // triangles. Thus, a detected collision doesn't mean that the point is
+  // actuallt inside the wrapped triangle and we have to check that here. This
+  // also gives us the opportunity to store the barycentric coordinates in case
+  // there is an intersection. Since the triangles don't overlap in this
+  // example, there is at most one triangle that contains a given point and we
+  // can abort the search early when we found a match.
+  template <typename Query, typename Primitive>
+  KOKKOS_FUNCTION auto operator()(Query const &query,
+                                  Primitive const &primitive) const
+  {
+    ArborX::ExperimentalHyperGeometry::Point<2> const &point =
+        getGeometry(getPredicate(query));
+    auto query_index = ArborX::getData(query);
+
+    auto const coeffs =
+        triangles_.get_mapping(primitive.index).get_coeff(point);
+    bool intersects = coeffs[0] >= 0 && coeffs[1] >= 0 && coeffs[2] >= 0;
+
+    if (intersects)
+    {
+      offsets_(query_index) = primitive.index;
+      coefficients_(query_index) = coeffs;
+      return ArborX::CallbackTreeTraversalControl::early_exit;
+    }
+    return ArborX::CallbackTreeTraversalControl::normal_continuation;
+  }
+
+private:
+  Triangles<DeviceType> triangles_;
+  Kokkos::View<int *, DeviceType> offsets_;
+  Kokkos::View<ArborX::Point *, DeviceType> coefficients_;
+};
+
+// Now that we have encapsulated the objects and queries to be used within the
+// Triangles class, we can continue with performing the actual search.
+int main()
+{
+  Kokkos::initialize();
+  {
+    using ExecutionSpace = Kokkos::DefaultExecutionSpace;
+    using MemorySpace = typename ExecutionSpace::memory_space;
+    using DeviceType = Kokkos::Device<ExecutionSpace, MemorySpace>;
+    ExecutionSpace execution_space;
+
+    std::cout << "Create grid with triangles.\n";
+    Triangles<DeviceType> triangles(execution_space);
+    std::cout << "Triangles set up.\n";
+
+    std::cout << "Creating BVH tree.\n";
+    ArborX::BasicBoundingVolumeHierarchy<
+        MemorySpace, ArborX::Details::PairIndexVolume<
+                         ArborX::ExperimentalHyperGeometry::Box<2>>> const
+        tree(execution_space, triangles);
+    std::cout << "BVH tree set up.\n";
+
+    std::cout << "Create the points used for queries.\n";
+    Points<DeviceType> points(execution_space);
+    std::cout << "Points for queries set up.\n";
+
+    std::cout << "Starting the queries.\n";
+    int const n = points.size();
+    Kokkos::View<int *, MemorySpace> offsets("offsets", n);
+    Kokkos::View<ArborX::Point *, MemorySpace> coefficients("coefficients", n);
```

Comment:
How do you distinguish whether intersection was found or not? The default values for these views are indistinguishable for the wrong ones.

Reply:
In `Debug` mode we run a full test checking that we detected the expected intersections and get the correct coefficients.